pydantic/schema.py

```diff
@@ -0,0 +1,392 @@
+from datetime import date, datetime, time, timedelta
+from decimal import Decimal
+from enum import Enum
+from pathlib import Path
+from typing import Any, Dict, Sequence, Set, Tuple, Type
+from uuid import UUID
+
+from . import main
+from .fields import Field, Shape
+from .json import pydantic_encoder
+from .types import DSN, UUID1, UUID3, UUID4, UUID5, DirectoryPath, EmailStr, FilePath, Json, NameEmail, UrlStr
+from .utils import clean_docstring
+
+
+def get_flat_models_from_model(model: Type['main.BaseModel']) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    assert issubclass(model, main.BaseModel)
+    flat_models.add(model)
+    for field in model.__fields__.values():
+        flat_models = flat_models | get_flat_models_from_field(field)
+    return flat_models
+
+
+def get_flat_models_from_field(field: Field) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    if field.sub_fields:
+        flat_models = flat_models | get_flat_models_from_sub_fields(field.sub_fields)
+    elif isinstance(field.type_, type) and issubclass(field.type_, main.BaseModel):
+        flat_models = flat_models | get_flat_models_from_model(field.type_)
+    return flat_models
+
+
+def get_flat_models_from_sub_fields(fields) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    for field in fields:
+        flat_models = flat_models | get_flat_models_from_field(field)
+    return flat_models
+
+
+def get_flat_models_from_models(models: Sequence[Type['main.BaseModel']]) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    for model in models:
+        flat_models = flat_models | get_flat_models_from_model(model)
+    return flat_models
+
+
+def get_long_model_name(model: Type['main.BaseModel']):
+    assert issubclass(model, main.BaseModel)
+    prefix = model.__module__.replace('.', '__')
+    name = model.__name__
+    return f'{prefix}__{name}'
+
+
+def get_model_name_maps(
+    unique_models: Set[Type['main.BaseModel']]
+) -> Tuple[Dict[str, Type['main.BaseModel']], Dict[Type['main.BaseModel'], str]]:
+    name_model_map: Dict[str, Type[main.BaseModel]] = {}
+    conflicting_names: Set[str] = set()
+    for model in unique_models:
+        model_name = model.__name__
+        if model_name in conflicting_names:
+            model_name = get_long_model_name(model)
+            name_model_map[model_name] = model
+        elif model_name in name_model_map:
+            conflicting_names.add(model_name)
+            conflicting_model = name_model_map[model_name]
+            del name_model_map[model_name]
+            conflicting_model_name = get_long_model_name(conflicting_model)
+            name_model_map[conflicting_model_name] = conflicting_model
+            model_name = get_long_model_name(model)
+            name_model_map[model_name] = model
```

Comment:
again this can be made more concise by combining to

```py
name_model_map[get_long_model_name(conflicting_model)] = conflicting_model
name_model_map[get_long_model_name(model)] = model
```

Reply:
Done